### DISCUSSION TITLE
Concurrecy not work for push

### DISCUSSION THREAD

**Discussion by Yiyiyimu** [community/community#27100] [Actions]
<p>Hi community,</p>
<p>When I tried to use the new feature concurrency in Github Actions, I find I don’t know how to choose the appropriate field for <code>group</code>. The example provided from documents is <code>${{ github.head_ref }}</code> which only works for or. For actions triggered by both push and or, I find a common choice is <code>${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}</code> from <a href="https://github.com/Project-MONAI/MONAI/pull/2285/files" rel="noopener nofollow ugc">MONAI</a> and <a href="https://github.com/apache/airflow/pull/15944/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR71-R74" rel="noopener nofollow ugc">airflow</a>, but unfortunately it is not working in my test: <a href="https://github.com/Yiyiyimu/GithubActionsTest/pull/8" class="inline-onebox" rel="noopener nofollow ugc">test: concurrency by Yiyiyimu · Pull Request #8 · Yiyiyimu/GithubActionsTest · GitHub</a></p>
<p>Is there any other recommendation on what to fill in <code>group</code>?</p>

**Comment by otter-computer** [comment]
<p>It depends exactly what level of concurrency you want. You could use <code>github.workflow</code> to have only one instance of the workflow running at a time, regardless of which branch triggered it, or even <code>github.repository</code> to only have one workflow running at a time inside the whole repository.</p>
<p>Can you give some more specifics on exactly what you’re aiming for when setting this up? That way we might be able to suggest a better option for you to use.</p>

**Comment by Yiyiyimu** [comment]
<p>Hi Thomas thanks for your prompt reply and glad to know the supported function of concurrency!</p>
<p>Actually what I want is to limit it to branch level. For example, if one action would be triggered by both push and pull_request, I want only two workflows running for this branch/PR, one for push and one for pr, and the latest commit would cancel the workflows triggered by previous ones. How should I implement this?</p>

**Comment by Simran-B** [comment, marked as answer]
<p>The last setup you tried looks like it should work and do what you want.</p>
<pre><code class="lang-yaml">name: test

on: [push, pull_request]

concurrency:
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
  cancel-in-progress: true
</code></pre>
<p>I would expect a second push to the same branch to cancel an in-flight workflow run triggered by a push.</p>
<p>Do I understand correctly, that both workflows are executed in parallel and both succeed instead of one being canceled and the other succeeding?</p>

**Comment by Yiyiyimu** [comment]
<p>Thank you Simran it solved the problem! It seems I forgot to make another push to test it :rofl:</p>

**Comment by chingc** [comment]
<p>I’d like to do the same thing as <a class="mention" href="/yiyiyimu">@Yiyiyimu</a> but I want the concurrency setting to apply to every branch EXCEPT the master branch.  Is there a way to do this?</p>

**Comment by Simran-B** [comment]
<p>I guess something like this should work:</p>
<pre><code class="lang-yaml">concurrency:
  group: ${{ github.workflow }}-${{ github.ref == 'refs/heads/master' &amp;&amp; github.run_number || github.ref }}
  cancel-in-progress: true
</code></pre>
<p>If it’s a push to the master branch, then it uses the run number which is unique (except if you re-run a workflow), so it shouldn’t cancel other concurrent runs. If it’s not the master branch, then only a single run per branch is allowed.</p>
<p>You could probably use github.run_id or github.sha instead of github.run_number with the same result.</p>

**Comment by laurie115** [comment]
<p>i hope your problem was solved</p>
